src/lightning_app/utilities/commands/base.py

```diff
@@ -0,0 +1,170 @@
+    except OSError as e:
+        if e.errno != errno.EEXIST and osp.isdir(path):
+            raise e
+
+
+class _Config(BaseModel):
+    command: str
+    affiliation: str
+    params: Dict[str, str]
+    is_command: bool
+    cls_path: str
+    cls_name: str
+    owner: str
+    requirements: Optional[List[str]]
+
+
+class ClientCommand:
+    def __init__(self, method: Callable, requirements: Optional[List[str]] = None) -> None:
+        self.method = method
+        flow = getattr(method, "__self__", None)
+        self.owner = flow.name if flow else None
+        self.requirements = requirements
+        self.metadata = None
+        self.models = Optional[Dict[str, BaseModel]]
+        self.app_url = None
+
+    def _setup(self, metadata: Dict[str, Any], models: Dict[str, BaseModel], app_url: str) -> None:
+        self.metadata = metadata
+        self.models = models
+        self.app_url = app_url
+
+    def run(self, **cli_kwargs) -> None:
+        """Overrides with the logic to execute on the client side."""
+
+    def invoke_handler(self, **kwargs: Any) -> Dict[str, Any]:
+        from lightning.app.utilities.state import headers_for
+
+        assert kwargs.keys() == self.models.keys()
+        for k, v in kwargs.items():
+            assert isinstance(v, self.models[k])
+        json = {
+            "command_name": self.metadata["command"],
+            "command_arguments": {k: v.json() for k, v in kwargs.items()},
+            "affiliation": self.metadata["affiliation"],
+            "id": str(uuid4()),
+        }
+        resp = requests.post(self.app_url + "/api/v1/commands", json=json, headers=headers_for({}))
+        assert resp.status_code == 200, resp.json()
+        return resp.json()
+
+    def _to_dict(self):
+        return {"owner": self.owner, "requirements": self.requirements}
+
+    def __call__(self, **kwargs: Any) -> Any:
+        assert self.models
+        kwargs = {k: self.models[k].parse_raw(v) for k, v in kwargs.items()}
+        return self.method(**kwargs)
+
+
+def _download_command(
+    command_metadata: Dict[str, Any], app_id: Optional[str]
+) -> Tuple[ClientCommand, Dict[str, BaseModel]]:
+    config = _Config(**command_metadata)
+    tmpdir = osp.join(gettempdir(), f"{getuser()}_commands")
+    makedirs(tmpdir)
+    target_file = osp.join(tmpdir, f"{config.command}.py")
+    if app_id:
+        client = LightningClient()
+        project_id = _get_project(client).project_id
+        response = client.lightningapp_instance_service_list_lightningapp_instance_artifacts(project_id, app_id)
+        for artifact in response.artifacts:
+            if f"commands/{config.command}.py" == artifact.filename:
+                r = requests.get(artifact.url, allow_redirects=True)
+                with open(target_file, "wb") as f:
+                    f.write(r.content)
+    else:
+        shutil.copy(config.cls_path, target_file)
+
+    cls_name = config.cls_name
```

Comment:
We 100% need to provide a way for users to inspect the downloaded file before importing, opt out of the process etc.
In general, the production-grade way of doing this will be installing a pip package.

Reply:
Yes, this needs to be scoped.